client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registration/RegistrationScreenModel.kt

```diff
@@ -0,0 +1,79 @@
+package presentation.auth.signup.registration
+
+import cafe.adriel.voyager.core.model.coroutineScope
+import domain.usecase.validation.IValidationUseCase
+import domain.utils.AuthorizationException
+import kotlinx.coroutines.CoroutineScope
+import presentation.base.BaseScreenModel
+
+class RegistrationScreenModel(private val validation: IValidationUseCase) :
+    BaseScreenModel<RegistrationUIState, RegistrationScreenEffect>(RegistrationUIState()),
+    RegistrationInteractionListener {
+    override val viewModelScope: CoroutineScope = coroutineScope
+
+    // region interactions
+    override fun onUsernameChanged(username: String) {
+        updateState { it.copy(username = username) }
+        tryCatch {
+            validation.validateUsername(username)
+            clearErrors()
+        }
+    }
+
+    override fun onPasswordChanged(password: String) {
+        updateState { it.copy(password = password) }
+        tryCatch {
+            validation.validatePassword(password)
+            clearErrors()
+        }
+    }
+
+    override fun onNextButtonClicked() {
+        with(state.value) {
+            tryCatch {
+                validation.validateUsername(username); validation.validatePassword(password)
+                sendNewEffect(
+                    RegistrationScreenEffect.NavigateToSubmitRegistrationScreen(username, password)
+                )
+            }
+        }
+    }
+
+    override fun onBackButtonClicked() {
+        sendNewEffect(RegistrationScreenEffect.NavigateBack)
+    }
+    // endregion
+
+    // region private methods
+    private fun tryCatch(block: () -> Unit) {
+        try {
+            block()
+        } catch (e: AuthorizationException.InvalidUsernameException) {
+            updateState {
+                it.copy(
+                    usernameErrorMsg = e.message ?: "Invalid Username",
```

Comment:
but how will localize this msg with this way?

Reply:
here is local validation not server side validation